### PR TITLE
review feat(role): field annotation with a role

### PR DESCRIPTION
**Pull request by tdurieux**
for each field of spoon defines for which interface it is the property

This PR is not finish I would like to have your feedback.

### CODE REVIEW

File: src/main/java/spoon/support/reflect/code/CtAssignmentImpl.java [hunk 1]
```diff
@@ -33,12 +34,16 @@
 public class CtAssignmentImpl<T, A extends T> extends CtStatementImpl implements CtAssignment<T, A> {
 	private static final long serialVersionUID = 1L;
 
+	@SpoonProperty(clazz = CtAssignment.class, name = SpoonProperty.PropertyName.ASSIGNED)
 	CtExpression<T> assigned;
 
+	@SpoonProperty(clazz = CtAssignment.class, name = SpoonProperty.PropertyName.ASSIGNMENT)
 	CtExpression<A> assignment;
 
+	@SpoonProperty(clazz = CtTypeReference.class, name = SpoonProperty.PropertyName.TYPE)
```
Comment (surli) [inline src/main/java/spoon/support/reflect/code/CtAssignmentImpl.java]: CtTypedElement

**Comment by pvojtechovsky** [comment]
I did not understood the purpose. What can profit from that?

The mapping of fields to correct property names, might make sense. But may be optional solution would be to rename fields, whose name do not fit to their property name. Then we might have simple rule: field name == property name (I have not checked if it is possible in all cases)

The mapping of property to interface `SpoonProperty#clazz` is ambiguous.
A) remove it completely (or let me understand what it is good for)
B) use an array of interfaces where this property occurs.

**Comment by pvojtechovsky** [comment]
Did I understood well, that this PR is a step on the way to the meta model of spoon model? 

In such case I suggest something like this:

C1) we need a `property descriptor` for each spoon model property. Properties are e.g. `CtTypedElement#type` or `CtNamedElement#name`
C2) the `property descriptor` knows interface(s?) on which this property is defined. Or it has at least method `#hasProperty(CtElement target) boolean` which returns `true` if the target element has  property defined by this `property descriptor`.
C3) the `property descriptor` knows type of the property. E.g. `List` or 'Map' or 'Set' or `single value` 
C4) for each type from C3, the property descriptor is implemented as sub class/interface of `property descritor`. So we will have 'list property descriptor', 'map property descriptor', ..., 'single value property descriptor'
C5) each `property descriptor` defines spoon model accessing methods. For example `single value property descriptor` would have accessors: `#getPropertyValue(CtElement target)` and `#setPropertyValue(CtElement target, Object value)`

I1) These property descriptors might be accessible as attribute of SpoonProperty.PropertyName enum. For example `SpoonProperty.PropertyName.DECLARING_TYPE.getDescriptor()` would return `property descriptor`, which might be used to access meta information about this property and to read/modify appropriate field of the spoon model.
I2) We need a CtScannerWithContext, which would provide access to `property descriptor` of just scanned property
...

**Comment by monperrus** [comment]
> Did I understood well, that this PR is a step on the way to the meta model of spoon model?

yes

> C1) we need a property descriptor for each spoon model property. Properties are e.g. CtTypedElement#type or CtNamedElement#name
> C3) the property descriptor knows type of the property. E.g. List or 'Map' or 'Set' or single value

Yes, this will be implemented as additional information in the to-be-renamed SpoonProperty.PropertyName enumeration (=I1)

> C2) the property descriptor knows interface(s?) on which this property is defined. 

Yes, this is the final goal

> Or it has at least method #hasProperty(CtElement target) boolean which returns true if the target element has property defined by this property descriptor.

Runtime support is welcome in subsequent PRs

> C5) each property descriptor defines spoon model accessing methods. For example single value property descriptor would have accessors: #getPropertyValue(CtElement target) and #setPropertyValue(CtElement target, Object value)

Yes, we aim at this too.

>  I2) We need a CtScannerWithContext, which would provide access to property descriptor of just scanned property

Welcome in subsequent PR.

It seems we're on the same wavelength.

**Comment by monperrus** [comment]
per our discussion:
```
@SpoonProperty -> @MetamodelProperty
SpoonProperty.PropertyName. -> reuse and rename CtPathRole -> CtRole
name = SpoonProperty.PropertyName.LEFT_OPERAND) -> role = CtRole.LEFT_OPERAND) 
```

contract: all non-final fields must be annotated with @MetamodelProperty (checkable with architecture test)

todo: remove the `clazz` information, create `@PropertyGetter(role = X),  @PropertySetter(role = X)`, in another PR that will be merged first (for this PR: contract 1): all roles are declared at least once in a metamodel interface method) and maybe another PR for contract 2): all methods of metamodel interfaces are annotated with either `@Unsettable, @DerivedProperty @PropertyGetter or @PropertySetter`

**Comment by tdurieux** [comment]
I updated this PR but I cannot create the contract because of these fields:

```java
private CtTypeReference<Void> voidType;
List<CtTypeParameterReference> formalTypeParameters = CtElementImpl.emptyList();
transient Factory factory;
protected CtElement parent;
SourcePosition position = SourcePosition.NOPOSITION;
Map<String, Object> metadata;
private CtMethod<T> valueOfMethod;
```

**Comment by tdurieux** [comment]
@monperrus Done for me

File: src/test/java/spoon/test/api/MetamodelTest.java [hunk 2]
```diff
@@ -46,4 +54,42 @@ public void testGetterSetterFroRole() {
 		Assert.assertEquals(expectedRoles, getterRoles);
 		Assert.assertEquals(expectedRoles, setterRoles);
 	}
+
+
+	@Test
+	/**
+	 * contract: all  all non-final fields must be annotated with {@link spoon.reflect.annotations.MetamodelPropertyField}
```
Comment (surli) [inline src/test/java/spoon/test/api/MetamodelTest.java]: Fix comment

**Comment by surli** [comment]
It should be great to have a test to check if all there is a getter/setter in one interface of the class for all field with a MetamodelPopertyRole.

**Comment by tdurieux** [comment]
Done

**Pull request merged**